File: golang/160-intersection-of-two-linked-lists.go
package golang

type ListNode struct {
	Val  int
	Next *ListNode
}

func getLength(node *ListNode) int {
    len := 0
    for node != nil {
        len++
        node = node.Next
    }
    return len
}

func shiftPointer(node *ListNode, k int) *ListNode {
    for k > 0 {
        node = node.Next
        k--
    }
    return node
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getIntersectionNode(headA, headB *ListNode) *ListNode {
    lengthA, lengthB := getLength(headA), getLength(headB)
    pointerA, pointerB := headA, headB
    if lengthA > lengthB {
        pointerA = shiftPointer(pointerA, lengthA - lengthB)
    }
    if lengthA < lengthB {
        pointerB = shiftPointer(pointerB, lengthB - lengthA)
    }
    for {
        if pointerA == pointerB {
            break
        }
        pointerA = pointerA.Next
        pointerB = pointerB.Next
    }
    return pointerA
}

File: javascript/160-intersection-of-two-linked-lists.js
function getLength(listNode) {
    let len = 0
    while (listNode !== null) {
        len++
        listNode = listNode.next
    }
    return len
}

function shiftPointer(listNode, k) {
    let p = listNode
    while (k-- > 0) {
        p = p.next
    }
    return p
}

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {  
    const [lengthA, lengthB] = [getLength(headA), getLength(headB)]
    let [pA, pB] = [headA, headB]
    if (lengthA > lengthB) {
        pA = shiftPointer(pA, Math.abs(lengthA - lengthB))
    }
    if (lengthA < lengthB) {
        pB = shiftPointer(pB, Math.abs(lengthA - lengthB))
    }
    while (true) {
        if (pA === pB) {
            return pA
        } else {
            pA = pA.next
            pB = pB.next
        }
    }
};